lib/puppet/provider/package/pip.rb

```diff
@@ -166,38 +202,67 @@ def latest_with_old_pip
       execpipe command_and_options do |process|
         process.collect do |line|
           # PIP OUTPUT: Using version 0.10.1 (newest of versions: 1.2.3, 4.5.6)
-          if line =~ /Using version (.+?) \(newest of versions/
-            return $1
+          if line =~ /Using version .+? \(newest of versions: (.+?)\)/
+            textAfterLastMatch = $1.chomp(")\n")
+            versionList = textAfterLastMatch.split(', ').sort do |x,y|
+              Puppet::Util::Package.versioncmp(x, y)
+            end
+            return versionList
           end
         end
         return nil
       end
     end
   end
 
+  # Finds the most suitable version available in a given range
+  def best_version(should_range)
+    included_available_versions = []
+    available_versions.each do |version|
+      version = PIP_VERSION.parse(version)
+      included_available_versions.push(version) if should_range.include?(version)
+    end
+
+    sorted_result = included_available_versions.sort { |x,y| x <=> y }
+    sorted_result.last
+  end
+
   # Install a package.  The ensure parameter may specify installed,
   # latest, a version number, or, in conjunction with the source
   # parameter, an SCM revision.  In that case, the source parameter
   # gives the fully-qualified URL to the repository.
-
   def install
     command = resource_or_provider_command
     self.class.validate_command(command)
 
+    should = @resource[:ensure]
     command_options = %w{install -q}
     command_options +=  install_options if @resource[:install_options]
     if @resource[:source]
-      if String === @resource[:ensure]
-        command_options << "#{@resource[:source]}@#{@resource[:ensure]}#egg=#{@resource[:name]}"
+      if String === should
+        command_options << "#{@resource[:source]}@#{should}#egg=#{@resource[:name]}"
       else
         command_options << "#{@resource[:source]}#egg=#{@resource[:name]}"
       end
     else
-      case @resource[:ensure]
-      when String
-        command_options << "#{@resource[:name]}==#{@resource[:ensure]}"
+      case should
       when :latest
         command_options << "--upgrade" << @resource[:name]
+      when String
+        begin
+          should_range = PIP_VERSION_RANGE.parse(should, PIP_VERSION)
+          should = best_version(should_range)
+
+          if should
+            command_options << "#{@resource[:name]}==#{should}"
+            @resource[:ensure] = should.to_s
```

Comment:
After discussing with @gimmyxd, we think it's worth mentioning the role of this line of code. When ensuring the installation of a python package using version ranges, the output would normally be something like:
> Notice: /Stage[main]/Main/Package[Flask]/ensure: ensure changed '0.12.5' to _**'>1'**_

But by adding this line of code, the same output becomes:
> Notice: /Stage[main]/Main/Package[Flask]/ensure: ensure changed '0.12.5' to _**'1.1.1'**_

The output is given by the `execute` method which uses said `@resource[:ensure]` variable.
@puppetlabs/night-s-watch: I personally found this more useful (from  user perspective) than seeing the version range given as you don't need to run additional commands, besides the puppet run, to find out what version was installed. There shouldn't be any impact on the existing functionality but feel free to address any concerns here.

Reply:
After chatting a bit on this topic, it was decided to remove this line until community/UX feedback is received.